app/services/designation-service.js

```diff
@@ -171,6 +176,108 @@ export default ApiBaseService.extend(NavigationAwareness, {
     return this.PUT(url, params);
   },
 
+  async resolveOrder(order) {
+    if (order) {
+      return order;
+    }
+    return this.userPickOrder({
+      state: ACTIVE_ORDER_STATES.join(",")
+    });
+  },
+
+  beginDesignation(options) {
+    let pkg = options.pkg,
+      order = options.order;
+
+    this.set("allowItemChange", options.allowItemChange);
+    this.set("allowOrderChange", options.allowOrderChange);
+
+    if (pkg.get("availableQuantity") > 0 || options.isEditOperation) {
+      this.triggerDesignateAction(pkg, order);
+    } else {
+      this.set("designationTargetPackage", pkg);
+      this.set("displayDesignateForm", true);
+      this.set("showItemAvailability", true);
+    }
+  },
+
+  triggerDesignateAction(pkg, order) {
+    this.resolveOrder(order).then(target => {
+      if (target) {
+        this.set("designationTargetPackage", pkg);
+        this.set("designationTargetOrder", target);
+        this.computeDesignationQuantities();
+        this.set("displayDesignateForm", true);
+        this.set("readyToDesignate", true);
+      }
+    });
+  },
+
+  triggerOrderChange(pkg, order) {
+    if (this.get("allowOrderChange")) {
+      this.triggerDesignateAction(pkg, order);
+    }
+  },
+
+  triggerItemChange() {
+    if (this.get("allowItemChange")) {
+      this.set("readyToDesignate", false);
+    }
+  },
+
+  computeDesignationQuantities() {
+    const pkg = this.get("designationTargetPackage");
+    const order = this.get("designationTargetOrder");
+
+    if (!order || !pkg) {
+      this.set("designatableQuantity", 0);
+      this.set("designationQty", 0);
+      return;
+    }
+
+    const maxQuantity =
+      pkg.get("availableQuantity") + this.alreadyDesignatedQuantity(pkg, order);
+
+    this.set("designatableQuantity", maxQuantity);
+    if (!this.get("designationQty")) {
+      this.set("designationQty", maxQuantity);
+    }
+  },
+
+  alreadyDesignatedQuantity(pkg, order) {
+    const ordPkg = pkg.get("ordersPackages").find(op => {
+      return parseInt(op.get("orderId")) === parseInt(order.get("id"));
+    });
+    return ordPkg ? ordPkg.get("quantity") : 0;
+  },
+
+  cancelDesignation() {
+    this.set("readyToDesignate", false);
+    this.set("designatableQuantity", 0);
+    this.set("designationTargetPackage", null);
+    this.set("designationTargetOrder", null);
+    this.set("designationQty", 0);
+  },
+
+  completeDesignation() {
+    if (!this.get("readyToDesignate")) return;
+
+    this.runTask(() => {
```

Comment:
runTask is async function

```suggestion
    await this.runTask(() => {
```